### PR TITLE
thirdpartyresource add multiple version support

### PR DESCRIPTION
**Pull request by adohe-zz**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36977)

### CODE REVIEW

File: pkg/api/meta/priority.go [hunk 1]
```diff
@@ -157,12 +158,19 @@ func (m PriorityRESTMapper) RESTMapping(gk unversioned.GroupKind, versions ...st
 	if err != nil {
 		return nil, err
 	}
-
 	// any versions the user provides take priority
 	priorities := m.KindPriority
 	if len(versions) > 0 {
 		priorities = make([]unversioned.GroupVersionKind, 0, len(m.KindPriority)+len(versions))
 		for _, version := range versions {
+			if strings.Count(version, "/") == 0 {
```
Comment (deads2k) [inline pkg/api/meta/priority.go]: This doesn't look right.  Doesn't `ParseGroupVersion` already do this? Also, open an issue to fix this method signature.  Either it gets a version or a groupversion, but we shouldn't have strings floating around anymore.

File: pkg/apiserver/api_installer.go [hunk 2]
```diff
@@ -108,6 +108,7 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
 	mediaTypes, streamMediaTypes := mediaTypesForSerializer(a.group.Serializer)
 	ws.Produces(append(mediaTypes, streamMediaTypes...)...)
 	ws.ApiVersion(a.group.GroupVersion.String())
+	ws.SetDynamicRoutes(true)
```
Comment (deads2k) [inline pkg/apiserver/api_installer.go]: This change isn't obvious to me.  Why would make this the new default?

File: pkg/apiserver/api_installer.go [hunk 3]
```diff
@@ -192,6 +193,7 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 	if err != nil {
 		return nil, err
 	}
+	fmt.Printf("register resource handler for path: %s and the fqKind is: %v\n", path, fqKindToRegister)
```
Comment (deads2k) [inline pkg/apiserver/api_installer.go]: debug

File: pkg/kubectl/resource/mapper.go [hunk 4]
```diff
@@ -54,12 +54,14 @@ func (m *Mapper) InfoForData(data []byte, source string) (*Info, error) {
 		return nil, fmt.Errorf("unable to decode %q: %v", source, err)
 	}
 
+	fmt.Printf("info for data gvk is: %v\n", gvk)
```
Comment (deads2k) [inline pkg/kubectl/resource/mapper.go]: debug

File: pkg/apiserver/api_installer.go (hunk 2 above)
Comment (AdoHe) [inline pkg/apiserver/api_installer.go]: Given our failed TPR integration test for example, if we create two TPR schemas: `foo.company.com` and `bar.company.com`, both with `v1`, thirdparty server will create a `WebService`, but if we delete `bar.company.com`, thirdparty server does nothing. And of course we cannot remove the entire web service, but we can remove `bar.company.com` related routes, as for go-restful, if we don't enable dynamic routes, we are note allowed to remove these routes.

File: pkg/apiserver/api_installer.go (hunk 3 above)
Comment (AdoHe) [inline pkg/apiserver/api_installer.go]: Done.

File: pkg/kubectl/resource/mapper.go (hunk 4 above)
Comment (AdoHe) [inline pkg/kubectl/resource/mapper.go]: Done.

File: pkg/api/meta/priority.go (hunk 1 above)
Comment (liggitt) [inline pkg/api/meta/priority.go]: v2 with empty group is not valid. All TPR would have `somegroup/v2`

File: pkg/apiserver/api_installer.go (hunk 2 above)
Comment (liggitt) [inline pkg/apiserver/api_installer.go]: I'm wary of changes that affect things other than the TPR endpoints. What if we made the resource endpoint a nested web service? Those were allowed to be added/removed previously without this setting

File: pkg/master/thirdparty/thirdparty.go [hunk 5]
```diff
@@ -100,23 +117,42 @@ func (m *ThirdPartyResourceServer) HasThirdPartyResource(rsrc *extensions.ThirdP
...
+		}
+		_, found := entry.storage[plural.Resource]
+		if !found {
+			return false, nil
+		}
+	}
+	return true, nil
 }
 
 func (m *ThirdPartyResourceServer) removeThirdPartyStorage(path, resource string) error {
```
Comment (liggitt) [inline pkg/master/thirdparty/thirdparty.go]: Can we change the signature so we get the info we need without string parsing?

File: pkg/master/thirdparty/thirdparty.go [hunk 6]
```diff
@@ -150,11 +189,38 @@ func (m *ThirdPartyResourceServer) RemoveThirdPartyResource(path string) error {
...
 		root := services[ix].RootPath()
 		if root == path || strings.HasPrefix(root, path+"/") {
-			m.genericAPIServer.HandlerContainer.Remove(services[ix])
+			ws = services[ix]
+			break
+		}
+	}
+
+	if ws != nil {
+		pattern := path + "/.*/" + resource + ".*"
```
Comment (liggitt) [inline pkg/master/thirdparty/thirdparty.go]: Why is the trailing `.*` needed

File: pkg/master/thirdparty/thirdparty.go (hunk 5 above)
Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: how about change this to `removeThirdPartyStorage(group, version, resource string)`, wdyt?

File: pkg/master/thirdparty/thirdparty.go (hunk 6 above)
Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: there should be some problem with this regexp, I am fixing it.

File: pkg/master/thirdparty/thirdparty.go (hunk 5 above)
Comment (liggitt) [inline pkg/master/thirdparty/thirdparty.go]: or `gvr unversioned.GroupVersionResource`

File: pkg/api/meta/priority.go (hunk 1 above)
Comment (AdoHe) [inline pkg/api/meta/priority.go]: the question is when version would be `group/version` or `version` when call restmapper?

Comment (deads2k) [inline pkg/api/meta/priority.go]: > @deads2k @liggitt I want to know why ParseGroupVersion use empty group when version is not group/version format, I think it should respect groupKind not use empty group.

`versions` should be just the version part.  Group should come from the `GroupKind`.

Comment (AdoHe) [inline pkg/api/meta/priority.go]: I will open a separate issue to describe this.

File: pkg/master/thirdparty/thirdparty.go (hunk 6 above)
Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: Done.

File: pkg/master/thirdparty/thirdparty.go (hunk 5 above)
Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: Done.

**Comment by AdoHe** [comment]
@k8s-bot unit test this

**Comment by AdoHe** [comment]
I get some time to get this updated, and fix all tests failure. @deads2k @liggitt @smarterclayton  please take a look, and give a deep review. I would like to get this merged and then go on with TPR finalizers.

File: pkg/apiserver/api_installer.go (hunk 2 above)
Comment (deads2k) [inline pkg/apiserver/api_installer.go]: > @liggitt this attribute is just webservice level, which gives us chance to remove routes handler of a specific webservice. I am not quite clear what you mean nested web service were allowed to be removed, would you mind explain more?

Can you create a webservice with a narrower scope for each TPR group?  This sort of change is why we're interested in breaking TPRs out of the main API server.

File: pkg/registry/extensions/rest/thirdparty_controller.go [hunk 7]
```diff
@@ -76,19 +78,31 @@ func (t *ThirdPartyController) SyncResources() error {
 }
 
 func (t *ThirdPartyController) syncResourceList(list runtime.Object) error {
-	existing := sets.String{}
+	existing := map[metav1.GroupVersionResource]sets.Empty{}
```
Comment (deads2k) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: `map[]struct{}`

File: pkg/registry/extensions/rest/thirdparty_controller.go [hunk 8]
```diff
@@ -76,19 +78,31 @@ func (t *ThirdPartyController) SyncResources() error {
...
 	switch list := list.(type) {
 	case *extensions.ThirdPartyResourceList:
 		// Loop across all schema objects for third party resources
 		for ix := range list.Items {
 			item := &list.Items[ix]
 			// extract the api group and resource kind from the schema
-			_, group, err := thirdpartyresourcedata.ExtractApiGroupAndKind(item)
+			kind, group, err := thirdpartyresourcedata.ExtractApiGroupAndKind(item)
```
Comment (deads2k) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: Seems like this API should return `GroupKind` and `GroupResource` for you.

File: pkg/registry/extensions/rest/thirdparty_controller.go [hunk 9]
```diff
@@ -18,28 +18,32 @@ package rest
...
 	// Install a third party resource described by 'rsrc'
 	InstallThirdPartyResource(rsrc *extensions.ThirdPartyResource) error
 	// Is a particular third party resource currently installed?
 	HasThirdPartyResource(rsrc *extensions.ThirdPartyResource) (bool, error)
-	// List all currently installed third party resources, the returned
-	// names are of the form <api-group-path>/<resource-plural-name>
-	ListThirdPartyResources() []string
+	// List all currently installed third party resources, return
+	// collection of all groupVersionResource
+	ListThirdPartyResources() []metav1.GroupVersionResource
```
Comment (deads2k) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: good change.

File: pkg/registry/extensions/rest/thirdparty_controller.go [hunk 10]
```diff
@@ -99,18 +113,18 @@ func (t *ThirdPartyController) syncResourceList(list runtime.Object) error {
...
+		for gvr := range existing {
+			if reflect.DeepEqual(installedGvr, gvr) {
 				found = true
 				break
 			}
 		}
 		// not expected, delete the resource
 		if !found {
-			if err := t.master.RemoveThirdPartyResource(installedAPI); err != nil {
+			if err := t.master.RemoveThirdPartyResource(installedGvr); err != nil {
```
Comment (deads2k) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: I see the fake for this changed, and the interface, but I can't seem to find where the actual impl changed.

File: pkg/registry/extensions/rest/thirdparty_controller.go (hunk 8 above)
Comment (AdoHe) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: I will check this when I return from ooo.

File: pkg/registry/extensions/rest/thirdparty_controller.go (hunk 10 above)
Comment (deads2k) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: nm, github helpfully hid the diff.

File: pkg/master/thirdparty/thirdparty.go [hunk 11]
```diff
@@ -84,12 +88,25 @@ func NewThirdPartyResourceServer(genericAPIServer *genericapiserver.GenericAPISe
 	return ret
 }
 
-// thirdPartyEntry combines objects storage and API group into one struct
+type thirdPartyGroup struct {
+	versionedEntry map[string]*thirdPartyEntry
```
Comment (deads2k) [inline pkg/master/thirdparty/thirdparty.go]: document the key for this map.  Version?

File: pkg/master/thirdparty/thirdparty.go [hunk 12]
```diff
@@ -84,12 +88,25 @@ func NewThirdPartyResourceServer(genericAPIServer *genericapiserver.GenericAPISe
...
+	group          metav1.APIGroup
+}
+
+// thirdPartyEntry combines objects storage into one struct
 // for easy lookup.
 type thirdPartyEntry struct {
```
Comment (deads2k) [inline pkg/master/thirdparty/thirdparty.go]: `tprVersionStorage`?

File: pkg/master/thirdparty/thirdparty.go [hunk 13]
```diff
@@ -84,12 +88,25 @@ func NewThirdPartyResourceServer(genericAPIServer *genericapiserver.GenericAPISe
...
-	// Map from plural resource name to entry
+	// Map from plural resource name to RESTStorage
 	storage map[string]*thirdpartyresourcedataetcd.REST
-	group   metav1.APIGroup
+}
+
+func (g *thirdPartyGroup) versionInstalled(version string) bool {
+	for installedVersion := range g.versionedEntry {
+		if reflect.DeepEqual(version, installedVersion) {
```
Comment (deads2k) [inline pkg/master/thirdparty/thirdparty.go]: deep equal on string is weird.

File: pkg/master/thirdparty/thirdparty.go [hunk 14]
```diff
@@ -60,8 +64,8 @@ type ThirdPartyResourceServer struct {
 
 	// storage for third party objects
 	thirdPartyStorageConfig *storagebackend.Config
-	// map from api path to a tuple of (storage for the objects, APIGroup)
-	thirdPartyResources map[string]*thirdPartyEntry
+	// map from api path to a tuple of (versioned storage for the objects, APIGroup)
+	thirdPartyResources map[string]*thirdPartyGroup
```
Comment (deads2k) [inline pkg/master/thirdparty/thirdparty.go]: keying by path is weird.  Why not key by group name and have the tprGroup hold a path value?

File: pkg/master/thirdparty/thirdparty.go [hunk 15]
```diff
@@ -101,61 +118,107 @@ func (m *ThirdPartyResourceServer) HasThirdPartyResource(rsrc *extensions.ThirdP
...
 		return false, nil
 	}
 	plural, _ := meta.KindToResource(schema.GroupVersionKind{
 		Group:   group,
 		Version: rsrc.Versions[0].Name,
 		Kind:    kind,
 	})
-	_, found := entry.storage[plural.Resource]
-	return found, nil
+	for _, version := range rsrc.Versions {
```
Comment (deads2k) [inline pkg/master/thirdparty/thirdparty.go]: Please rename `rsrc` to `tpr`.  I keep having to look up what this is.

File: pkg/master/thirdparty/thirdparty.go [hunk 16]
```diff
@@ -84,12 +88,25 @@ func NewThirdPartyResourceServer(genericAPIServer *genericapiserver.GenericAPISe
...
+func (g *thirdPartyGroup) versionInstalled(version string) bool {
+	for installedVersion := range g.versionedEntry {
+		if reflect.DeepEqual(version, installedVersion) {
+			return true
+		}
+	}
+	return false
 }
 
 // HasThirdPartyResource returns true if a particular third party resource currently installed.
```
Comment (deads2k) [inline pkg/master/thirdparty/thirdparty.go]: if its installed in one version and not another, what's the behavior?

Comment (deads2k) [inline pkg/master/thirdparty/thirdparty.go]: I think I actually expect this to take a GVR.  That would make it unambiguous.

File: pkg/master/thirdparty/thirdparty.go [hunk 17]
```diff
@@ -101,61 +118,107 @@ func (m *ThirdPartyResourceServer) HasThirdPartyResource(rsrc *extensions.ThirdP
...
 		root := services[ix].RootPath()
 		if root == path || strings.HasPrefix(root, path+"/") {
-			m.genericAPIServer.HandlerContainer.Remove(services[ix])
+			ws = services[ix]
+			break
+		}
+	}
+
+	if ws != nil {
+		pattern := path + "/.*" + gvr.Resource + ".*"
```
Comment (deads2k) [inline pkg/master/thirdparty/thirdparty.go]: why the wildcarding?  This looks like it would hit false positives.

**Comment by deads2k** [comment]
I think this may just be too big to fix inside of our current API server. The changes are pervasive and are changing attributes our current RESTHandler to allow removal.  I worry about accidentally removing something important.

@lavalamp @kubernetes/sig-api-machinery-misc If we ran TPRs out of a addon API server, we could have a completely separate handler chain and have confidence of containment for changes like this.

File: pkg/registry/extensions/rest/thirdparty_controller.go (hunk 7 above)
Comment (AdoHe) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: Done.

File: pkg/registry/extensions/rest/thirdparty_controller.go (hunk 8 above)
Comment (AdoHe) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: maybe we can return `GroupKind` here, seems we cannot extract `GroupResource` from item, wdyt?

File: pkg/registry/extensions/rest/thirdparty_controller.go (hunk 10 above)
Comment (AdoHe) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: actually, the implementation changed a lot :)

File: pkg/master/thirdparty/thirdparty.go (hunk 11 above)
Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: Done.

File: pkg/master/thirdparty/thirdparty.go (hunk 12 above)
Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: good suggestion :)

File: pkg/master/thirdparty/thirdparty.go (hunk 13 above)
Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: Done.

File: pkg/master/thirdparty/thirdparty.go (hunk 15 above)
Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: Done.

File: pkg/master/thirdparty/thirdparty.go [hunk 18]
```diff
@@ -101,61 +118,107 @@ func (m *ThirdPartyResourceServer) HasThirdPartyResource(rsrc *extensions.ThirdP
...
+	entry, found := tpg.versionedEntry[gvr.Version]
 	if !found {
 		return nil
 	}
-	storage, found := entry.storage[resource]
+	storage, found := entry.storage[gvr.Resource]
 	if !found {
 		return nil
 	}
 	if err := m.removeAllThirdPartyResources(storage); err != nil {
```
Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: actually it is deleting all TPR data in this storage registry.

File: pkg/master/thirdparty/thirdparty.go (hunk 17 above)
Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: yes, this is still not correct.

File: pkg/master/thirdparty/thirdparty.go (hunk 16 above)
Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: if we change `HasThirdPartyResource(rsrc)` to `HasThirdPartyResource(gvr)`, then I would suppose we should make the same change to `InstallThirdPartyResource`, which means we should update the thirdparty controller. wdyt?

Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: one TPR schema could define multiple `GroupVersionResource` identified resource, so update method to accept `gvr` parameter could make sense.

**Comment by AdoHe** [comment]
@lavalamp could you please take a look? If we decide to run TPRs out of a addon API server, I would start building this addon server, and add this to that server.

File: pkg/apiserver/api_installer.go (hunk 2 above)
Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: +1 for putting TPRs in a separate ws, so that there's no way of accidentally removing a route for a built-in API.

Comment (AdoHe) [inline pkg/apiserver/api_installer.go]: @lavalamp I will try to update this, but what I am really confused is whether I should go on this PR against current apiserver or I should start working on a add-on server?

Comment (zhouhaibing089) [inline pkg/apiserver/api_installer.go]: @AdoHe is there already an addon server in developing status?

Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: IMO it's not bad to make improvements to the existing TPR implementation. We won't be moving it this quarter, I'm pretty sure.

**Comment by bgrant0607** [comment]
See also #24299

**Comment by liggitt** [comment]
Since TPRs cannot be converted between versions by the server, we cannot enable multiple versions in the API server and behave like a standard Kubernetes API.

**Comment by AdoHe** [comment]
@liggitt yes, so I am going to close this and will re-implement it in an addon server.

**Comment by jamiehannaford** [comment]
@adohe Did you get a chance to start work on the addon server? I'm interested in helping out with TPRs, so was interested in the current status of things

**Comment by deads2k** [comment]
> @adohe Did you get a chance to start work on the addon server? I'm interested in helping out with TPRs, so was interested in the current status of things

We're looking at this proposal to move things forward: https://github.com/kubernetes/community/pull/524.

**Pull request closed without merging**